base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -22,97 +22,323 @@
+use monero::{
+    blockdata::{
+        block::BlockHeader as MoneroBlockHeader,
+        transaction::SubField,
+        Transaction as MoneroTransaction,
+        TransactionPrefix,
+    },
+    consensus::encode::VarInt,
+    cryptonote::hash::{Hash, Hashable},
+};
 #[cfg(feature = "monero_merge_mining")]
 use randomx_rs::{RandomXCache, RandomXDataset, RandomXError, RandomXFlag, RandomXVM};
 use serde::{Deserialize, Serialize};
-use tari_mmr::MerkleProof;
 
 const MAX_TARGET: U256 = U256::MAX;
+const SEEDHASH_EPOCH: u64 = 2048;
+const SEEDHASH_LAG: u64 = 64;
 
 #[derive(Debug, Error, Clone)]
 enum MergeMineError {
     // Error deserializing Monero data
     DeserializeError,
+    // Error serializing Monero data
+    SerializeError,
     // Hashing of Monero data failed
     HashingError,
+    // Validation Failure
+    ValidationError,
     // RandomX Failure
     #[cfg(feature = "monero_merge_mining")]
     RandomXError(RandomXError),
 }
 
 /// This is a struct to deserialize the data from he pow field into data required for the randomX Monero merged mine
 /// pow.
-#[derive(Serialize, Deserialize, Clone, Debug, Default)]
+#[derive(Serialize, Deserialize, Clone, Debug)]
 pub struct MoneroData {
     // Monero header fields
-    // #[serde(with = "HashMoneroHeader")]
-    header: MoneroBlockHeader,
+    pub header: MoneroBlockHeader,
     // randomX vm key
-    key: String,
+    pub seed_key: Vec<u8>,
+    // Monero block height
+    pub height: u64,
     // transaction count
-    count: u16,
+    pub count: u16,
     // transaction root
-    transaction_root: [u8; 32],
+    pub transaction_root: Vec<u8>,
     // Transaction proof of work.
-    merkle_proof: MerkleProof,
+    pub merkle_proof: MoneroMerkleProof,
     // Coinbase tx from Monero
-    coinbase_tx: MoneroTransaction,
+    pub coinbase_tx: MoneroTransaction,
 }
 
+// todo this needs database access
+// fn check_randomx_key(data: &MoneroData) -> Result<(), MergeMineError> {
+//     let age_limit = 4 * data.height * (data.block_time / 120); // get `120` from consensus constants instead
+//     if data.initial_use > (data.height - age_limit) {
+//         return Ok(());
+//     }
+//     Err(MergeMineError::ValidationError)
+// }
+
 impl MoneroData {
+    fn verify_merkle_root(&self) -> Result<(), MergeMineError> {
+        if !self.merkle_proof.validate(&self.transaction_root) {
+            return Err(MergeMineError::ValidationError);
+        }
+        let hash = self.coinbase_tx.hash().to_bytes();
+        if !self.merkle_proof.validate_value(&hash.to_vec()) {
+            return Err(MergeMineError::ValidationError);
+        }
+
+        Ok(())
+    }
+
     fn new(tari_header: &BlockHeader) -> Result<MoneroData, MergeMineError> {
         bincode::deserialize(&tari_header.pow.pow_data).map_err(|_| MergeMineError::DeserializeError)
     }
+
+    // I think this is provided by the merged miner
+
+    fn generate(tari_header: &mut BlockHeader) {
```

Comment:
I think this is provided by the merged miner. It should only be used for test. Perhaps we need to wrap in in cfg test

Reply:
Agreed, intention was to use it to generate test data.

Reply:
This was removed in #2061